File: geekbang/algorithm-models/2.1-前缀和、数组计数代码模板/numberOfSubarrays.py
# Python
# LeetCode 1248 统计优美子数组
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        nums = [0] + nums
        s = [0] * len(nums)
        for i in range(1, len(nums)):
            s[i] = s[i - 1] + nums[i] % 2
        count = [0] * len(s)
        for i in range(len(s)):
            count[s[i]] += 1
        ans = 0
        for i in range(1, len(s)):
            if s[i] - k >= 0:
                ans += count[s[i] - k]
        return ans

File: geekbang/algorithm-models/2.1-前缀和、数组计数代码模板/numberOfSubarrays.java
// Java
// LeetCode 1248 统计优美子数组
class Solution {
    public int numberOfSubarrays(int[] nums, int k) {
        int n = nums.length;
        int[] s = new int[n + 1]; // 0~n
        int[] count = new int[n + 1];
        // s[0] = 0;
        count[s[0]]++;
        for (int i = 1; i <= n; i++) {
            s[i] = s[i - 1] + nums[i - 1] % 2;
            count[s[i]]++;
        }


        int ans = 0;
        for (int i = 1; i <= n; i++) {
            // s[i] - s[j] = k, 求j的数量
            // s[j] = s[i] - k
            if (s[i] - k >= 0) {
                ans += count[s[i] - k];
            }
        }
        return ans;
    }
}
